File: book2-algorithms-sedgewick/ch2/05_bottom_up_merge_sort.cpp
#include <iostream>
#include <algorithm>
#include <utility>
#include <random>
#include <string>
#include <vector>
#include <chrono>

template<typename T>
class BottomUpMergeSort {
public:

    BottomUpMergeSort() {};

    void sort(T& arr) {
        auto len = arr.size();
        T aux;
        aux.resize(len);
        for (size_t sz = 1; sz < len; sz = sz + sz)
            for (size_t lo = 0; lo < len - sz; lo += 2 * sz)
                merge(aux, arr, lo, lo + sz - 1, std::min(lo + sz + sz - 1, len - 1));
    }

private:
    void merge(T& aux, T& arr, size_t left, size_t mid, size_t right) {
        // copy original data into aux array.
        for (auto idx = left; idx <= right; idx++)
            aux[idx] = arr[idx];
        // merge in arr
        auto i = left;
        auto j = mid + 1;
        for (auto k = left; k <= right; k++) {
            if (i > mid) arr[k] = aux[j++]; // left subarray exhausted, take right subarray.
            else if (j > right) arr[k] = aux[i++]; // right subarray exhausted, take left subarray.
            else if (aux[j] < aux[i]) arr[k] = aux[j++];
            else arr[k] = aux[i++];
        }
    }
};

int main(int argc, char* argv[]) {
    using std::chrono::high_resolution_clock;
    using std::chrono::duration_cast;
    using std::chrono::milliseconds;

    std::cout << "Bottom Up Merge Sort." << std::endl;

    if (argc <= 1) {
        std::cerr << "Missing number of elements." << std::endl;
        return EXIT_FAILURE;
    }

    size_t numbers_length = std::stoi(argv[1]);
    std::cout << "Number of Elements: " << numbers_length << std::endl;

    std::cout << "Generating data... " << std::flush;
    std::vector<long> numbers;
    std::random_device rd;
    std::mt19937 gen(rd());
    std::uniform_int_distribution <size_t> dist(0, 1000000);
    while (numbers_length > 0) {
        numbers.push_back(dist(gen));
        numbers_length--;
    }
    std::cout << "Done." << std::endl;

    if (numbers.size() < 100) {
        auto it = numbers.begin();
        std::cout << "Numbers: ";
        while (it != numbers.end()) {
            std::cout << *it << " ";
            it++;
        }
        std::cout << std::endl;
    }

    std::cout << "Sorting... " << std::flush;
    auto t1 = high_resolution_clock::now();
    BottomUpMergeSort<std::vector<long>> bottom_up_merge_sort;
    bottom_up_merge_sort.sort(numbers);
    auto t2 = high_resolution_clock::now();
    std::cout << "Done." << std::endl;
    auto running_time = duration_cast<milliseconds>(t2 - t1).count();
    std::cout << "Running Time: " << running_time << "ms" << std::endl;

    if (numbers.size() < 100) {
        auto it = numbers.begin();
        std::cout << "Numbers: ";
        while (it != numbers.end()) {
            std::cout << *it << " ";
            it++;
        }
        std::cout << std::endl;
    }

    std::cout << "Checking... " << std::flush;
    for (size_t i = 0; i < numbers.size() - 1; i++) {
        if (numbers[i] > numbers[i + 1]) {
            std::cerr << "Bad sorting: " << numbers[i] << ", " << numbers[i + 1] << std::endl;
        }
    }
    std::cout << "Done." << std::endl;

    return EXIT_SUCCESS;
}

File: book2-algorithms-sedgewick/ch2/05_bottom_up_merge_sort.py
from math import floor
import sys
import random
import time

class BottomUpMergeSort:

    def sort(self, items) -> None:
        aux = items[:]
        sz = 1
        while sz < len(items):
            for lo in range(0, len(items) - sz, 2 * sz):
                self._merge(items, aux, lo, lo + sz - 1, min(lo + 2 * sz - 1, len(items) - 1))
            sz += sz

    def _merge(self, items, aux, left, mid, right) -> None:
        # copy elements to aux.
        for i in range(left, right + 1):
            aux[i] = items[i]
        
        # merge both subarrays.
        i = left
        j = mid + 1
        for k in range(left, right + 1):
            if i > mid:
                items[k] = aux[j]
                j += 1
            elif j > right:
                items[k] = aux[i]
                i += 1
            elif aux[i] > aux[j]:
                items[k] = aux[j]
                j += 1
            else:
                items[k] = aux[i]
                i += 1
            k += 1


def main():
    print('Bottom Up Merge Sort')

    if len(sys.argv) < 2:
        print('Missing number of items.')
        exit(1)
    
    number_items = int(sys.argv[1])
    print('Number of Items: ', number_items)

    print('Generating numbers... ', end='', flush=True)
    numbers = [random.randint(0, 1000000) for _ in range(number_items)]
    print('Done.')

    if number_items <= 100:
        print('Numbers: ', ', '.join((str(n) for n in numbers)))
    
    print('Sorting... ', end='', flush=True)
    bottom_up_merge_sort = BottomUpMergeSort()
    t1 = time.time()
    bottom_up_merge_sort.sort(numbers)
    t2 = time.time()
    print('Done.')

    print('Running time: ', (t2 - t1), 'secs.')

    if number_items <= 100:
        print('Numbers: ', ', '.join((str(n) for n in numbers)))

    print('Checking... ', end='', flush=True)
    for i in range(len(numbers) - 1):
        if numbers[i] > numbers[i + 1]:
            print('\nBad ordering: ', numbers[i], ', ', numbers[i + 1])
            exit(1)
    print('Done.')

if __name__ == '__main__':
    main()
